include/pybind11/detail/common.h

```diff
@@ -447,23 +446,33 @@
 PYBIND11_WARNING_PUSH
 PYBIND11_WARNING_DISABLE_CLANG("-Wgnu-zero-variadic-macro-arguments")
 #define PYBIND11_MODULE(name, variable, ...)                                                      \
-    static ::pybind11::module_::module_def PYBIND11_CONCAT(pybind11_module_def_, name)            \
-        PYBIND11_MAYBE_UNUSED;                                                                    \
-    PYBIND11_MAYBE_UNUSED                                                                         \
+    static ::pybind11::module_::module_def PYBIND11_CONCAT(pybind11_module_def_, name);           \
+    static std::vector<PyModuleDef_Slot> PYBIND11_CONCAT(pybind11_module_slots_, name);           \
```

Comment:
Slight worries: Could we get into trouble during process teardown?

Concretely, could there be a situation in which `slots.data()` is referenced after this `static` variable was deleted already?

Probably not ... but maybe?

I'd want to err on the side of "abundance of caution".

I think we should use the standard trick of leaking a pointer. It's really simple, all doubts removed. **Untested**:

```
    static auto *PYBIND11_CONCAT(pybind11_module_slots_, name) = new std::vector<PyModuleDef_Slot>();
```

Reply:
I'd prefer not to leak a pointer unless we are sure it is needed. Isn't this already the lifetime of the module, though?

Reply:
Yes, this is a static global for the lifetime of the module... it ends up referenced inside the initialized PyModuleDef, which is defined right above it.  So I don't think there's any way that the PyModuleDef could outlive it.

Reply:
It's standard best practice for non-POD `static` objects.

There is no good reason for taking any risks (running destructors during process teardown; or freeing in the wrong order). The memory will be freed by the kernel for sure.

Reply:
Ok, I switched it to an std::array of slots, which is POD (since PyModuleDef_Slot is POD), so it has no destructor.... it's more along the lines of the original suggestion of using a named constant instead of a magic number for the size.... the size is part of the typedef (with a comment, ofc).